logstash-core/lib/logstash/output_delegator.rb

```diff
@@ -1,192 +1,60 @@
+  def initialize(logger, output_class, metric, xopts={}, plugin_args={})
     @logger = logger
-    @threadsafe = klass.threadsafe?
-    @config = plugin_args.reduce({}, :merge)
-    @klass = klass
-    @workers = java.util.concurrent.CopyOnWriteArrayList.new
-    @default_worker_count = default_worker_count
-    @registered = false
-
-    # Create an instance of the input so we can fetch the identifier
-    output = @klass.new(@config)
-
-    # Scope the metrics to the plugin
-    namespaced_metric = metric.namespace(output.plugin_unique_name.to_sym)
-    output.metric = namespaced_metric
+    @output_class = output_class
+    @metric = metric
+    @xopts = xopts # Options specific to the given class
+    @plugin_args = plugin_args
+    build_strategy!
 
-    @metric_events = namespaced_metric.namespace(:events)
-    namespaced_metric.gauge(:name, config_name)
-
-    @events_received = Concurrent::AtomicFixnum.new(0)
+    @id = plugin_args["id"]
+    @namespaced_metric = metric.namespace("#{@output_class.config_name}_#{@id}".to_sym)
+    @metric_events = @namespaced_metric.namespace(:events)
+    @namespaced_metric.gauge(:name, config_name)
   end
 
-  def threadsafe?
-    !!@threadsafe
+  def concurrency
+    @output_class.concurrency
   end
 
-  def warn_on_worker_override!
-    # The user has configured extra workers, but this plugin doesn't support it :(
-    if worker_limits_overriden?
-      message = @klass.workers_not_supported_message
-      warning_meta = {:plugin => @klass.config_name, :worker_count => @config["workers"]}
-      if message
-        warning_meta[:message] = message
-        @logger.warn(I18n.t("logstash.pipeline.output-worker-unsupported-with-message", warning_meta))
-      else
-        @logger.warn(I18n.t("logstash.pipeline.output-worker-unsupported", warning_meta))
-      end
-    end
-  end
+  def build_strategy!
+    # TODO: Make this injected
+    @strategy_registry = LogStash::OutputDelegatorStrategyRegistry.instance
+    @strategy_class = @strategy_registry.get(self.concurrency)
 
-  def worker_limits_overriden?
-    @config["workers"] && @config["workers"] > 1 && @klass.workers_not_supported?
+    if !@strategy_class
+      raise "Could not find output delegator strategy of type '#{self.concurrency}'. Valid strategies:#{@strategy_registry.types}"
+    end
+    
+    @strategy = @strategy_class.new(@logger, @output_class, @metric, @xopts, @plugin_args)
   end
 
-  def target_worker_count
-    # Remove in 5.0 after all plugins upgraded to use class level declarations
-    raise ArgumentError, "Attempted to detect target worker count before instantiating a worker to test for legacy workers_not_supported!" if @workers.size == 0
-
-    if @threadsafe || @klass.workers_not_supported?
-      1
-    else
-      @config["workers"] || @default_worker_count
-    end
+  def id
+    @plugin_args["id"]
   end
 
   def config_name
-    @klass.config_name
+    "#{@output_class.config_name}_#{id}"
```

Comment:
I feel the same way, however I did not comment on this because it is part of the public API. But, if id is now an intrinsic part of the config_name then it is correct.